src_common/uint128.c

```diff
@@ -0,0 +1,292 @@
+    uint32_t diffBits = bits128(l) - bits128(r);
+    clear128(&resDiv);
+    copy128(&resMod, l);
+    if (gt128(r, l)) {
+        copy128(retMod, l);
+        clear128(retDiv);
+    } else {
+        shiftl128(r, diffBits, &copyd);
+        shiftl128(&one, diffBits, &adder);
+        if (gt128(&copyd, &resMod)) {
+            shiftr128(&copyd, 1, &copyd);
+            shiftr128(&adder, 1, &adder);
+        }
+        while (gte128(&resMod, r)) {
+            if (gte128(&resMod, &copyd)) {
+                sub128(&resMod, &copyd, &resMod);
+                or128(&resDiv, &adder, &resDiv);
+            }
+            shiftr128(&copyd, 1, &copyd);
+            shiftr128(&adder, 1, &adder);
+        }
+        copy128(retDiv, &resDiv);
+        copy128(retMod, &resMod);
+    }
+}
+
+bool tostring128(const uint128_t *const number,
+                 uint32_t baseParam,
+                 char *const out,
+                 uint32_t outLength) {
+    uint128_t rDiv;
+    uint128_t rMod;
+    uint128_t base;
+    copy128(&rDiv, number);
+    clear128(&rMod);
+    clear128(&base);
+    LOWER(base) = baseParam;
+    uint32_t offset = 0;
+    if ((baseParam < 2) || (baseParam > 16)) {
+        return false;
+    }
+    do {
+        if (offset > (outLength - 1)) {
+            return false;
+        }
+        divmod128(&rDiv, &base, &rDiv, &rMod);
+        out[offset++] = HEXDIGITS[(uint8_t) LOWER(rMod)];
+    } while (!zero128(&rDiv));
+
+    if (offset > (outLength - 1)) {
+        return false;
+    }
+
+    out[offset] = '\0';
+    reverseString(out, offset);
+    return true;
+}
+
+/**
+ * Format a uint128_t into a string as a signed integer
+ *
+ * @param[in] number the number to format
+ * @param[in] base the radix used in formatting
+ * @param[out] out the output buffer
+ * @param[in] out_length the length of the output buffer
+ * @return whether the formatting was successful or not
+ */
+bool tostring128_signed(const uint128_t *const number,
+                        uint32_t base,
+                        char *const out,
+                        uint32_t out_length) {
+    uint128_t max_unsigned_val;
+    uint128_t max_signed_val;
+    uint128_t one_val;
+    uint128_t two_val;
+    uint128_t tmp;
+
+    // showing negative numbers only really makes sense in base 10
+    if (base == 10) {
+        explicit_bzero(&one_val, sizeof(one_val));
+        LOWER(one_val) = 1;
+        explicit_bzero(&two_val, sizeof(two_val));
+        LOWER(two_val) = 2;
+
+        memset(&max_unsigned_val, 0xFF, sizeof(max_unsigned_val));
+        divmod128(&max_unsigned_val, &two_val, &max_signed_val, &tmp);
+        if (gt128(number, &max_signed_val))  // negative value
+        {
+            sub128(&max_unsigned_val, number, &tmp);
+            add128(&tmp, &one_val, &tmp);
+            out[0] = '-';
+            return tostring128(&tmp, base, out + 1, out_length - 1);
+        }
+    }
+    return tostring128(number, base, out, out_length);  // positive value
+}
```

Comment:
This file either needs to source the paper from which the formulas are from or a lot more comments that explains all the steps

Reply:
Out of the scope of this PR, this code was only split and isn't new.